bin/propolis-server/src/lib/stats/virtual_machine.rs

```diff
@@ -0,0 +1,526 @@
+        }
+    }
+
+    fn test_usage() -> VcpuUsage {
+        VcpuUsage {
+            state: "run".to_string(),
+            vcpu_id: 0,
+            datum: Cumulative::new(100),
+        }
+    }
+
+    #[test]
+    fn test_no_schema_changes() {
+        let mut set = SchemaSet::default();
+        assert!(set
+            .insert_checked(&test_virtual_machine(), &test_usage())
+            .is_none());
+        const PATH: &str = concat!(
+            env!("CARGO_MANIFEST_DIR"),
+            "/tests/output/virtual-machine-schema.json",
+        );
+        set.assert_contents(PATH);
+    }
+
+    #[test]
+    fn test_kstat_instance_from_instance_id() {
+        let ks = Kstat {
+            ks_module: VMM_KSTAT_MODULE_NAME,
+            ks_instance: 0,
+            ks_name: VM_KSTAT_NAME,
+            ks_snaptime: 1,
+        };
+        const INSTANCE_ID: &str = "db198b43-2dee-4b4b-8a68-24cb4c0d6ec8";
+        let data = Data::Named(vec![Named {
+            name: VM_NAME_KSTAT,
+            value: NamedData::String(INSTANCE_ID),
+        }]);
+
+        assert_eq!(
+            kstat_instance_from_instance_id(&ks, &data, INSTANCE_ID)
+                .expect("Should have matched the instance ID"),
+            ks.ks_instance,
+        );
+
+        let data = Data::Named(vec![Named {
+            name: VM_NAME_KSTAT,
+            value: NamedData::String("something-else"),
+        }]);
+        assert!(
+            kstat_instance_from_instance_id(&ks, &data, INSTANCE_ID).is_none(),
+            "Should not have matched an instance ID"
+        );
+    }
+
+    fn vcpu_state_kstats<'a>() -> (Kstat<'a>, Data<'a>) {
+        let ks = Kstat {
+            ks_module: VMM_KSTAT_MODULE_NAME,
+            ks_instance: 0,
+            ks_name: "vcpu0",
+            ks_snaptime: 1,
+        };
+        let data = Data::Named(vec![
+            Named { name: VCPU_KSTAT_PREFIX, value: NamedData::UInt32(0) },
+            // There are three ustates, but the first two are aggregated.
+            Named { name: "time_init", value: NamedData::UInt64(1) },
+            Named { name: "time_idle", value: NamedData::UInt64(1) },
+            Named { name: "time_run", value: NamedData::UInt64(2) },
+        ]);
+        (ks, data)
+    }
+
+    #[test]
+    fn test_produce_vcpu_usage() {
+        let (ks, data) = vcpu_state_kstats();
+        let kstats = [(Utc::now(), ks, data)];
+        let samples =
+            produce_vcpu_usage(&test_virtual_machine(), kstats.iter())
+                .expect("Should have produced samples");
+        assert_eq!(
+            samples.len(),
+            2,
+            "Should have samples for 'run' and 'idle' states"
+        );
+        for ((sample, state), x) in
+            samples.iter().zip(["idle", "run"]).zip([2, 2])
+        {
+            let st = sample
+                .fields()
+                .iter()
+                .find_map(|f| {
+                    if f.name == "state" {
```

Comment:
these state names are also used above, would it make sense for them to be constants?

Reply:
Made the state names all consts, and did a few other bits in 7c4cbe7